File: 0234-palindrome-linked-list/0234-palindrome-linked-list.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    /*
    1221
    21
    Algorithm                              
    1) copy store the element        OR 1)check if the begining and ending are equal
    2) then reverse the element         2) with 2ptrs until equal
    3) compare                          
     
    
    */
    bool isPalindrome(ListNode* head) {
      vector<int>ans;
        while(head){//while not NULL
            ans.push_back(head->val);//copy list to vector
            head=head->next;//increament list
        }
        int s=0,e=ans.size()-1;
        while(s<e){
            if(ans[s]!=ans[e])//compare numbers
                return false;
                s++;
                e--;
        }
        return true;
    }
};

File: 0234-palindrome-linked-list/0234-palindrome-linked-list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def isPalindrome(self, head: Optional[ListNode]) -> bool:
        check = True
        slow = head
        fast = head
        trail = head
        while fast and fast.next:
            fast = fast.next.next
            slow = slow.next
            
        prev = None
        curr = slow
        
        while curr:
            nxt = curr.next
            curr.next = prev
            prev = curr
            curr = nxt
        
        
        while prev and head:
            if head.val != prev.val:
                check = False
                break
                
            prev = prev.next
            head = head.next
            
        return check